File: problems/03-sliding-window/001-longest-substring-without-repeating-characters/java/Solution.java
import java.util.HashSet;
import java.util.Set;

public class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> window = new HashSet<>();
        int left = 0;
        int best = 0;

        for (int right = 0; right < s.length(); right++) {
            char current = s.charAt(right);

            while (window.contains(current)) {
                // Remove the leftmost character until the window is valid again.
                window.remove(s.charAt(left));
                left++;
            }

            window.add(current);
            best = Math.max(best, right - left + 1);
        }

        return best;
    }
}

File: problems/03-sliding-window/001-longest-substring-without-repeating-characters/python/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        window = set()
        left = 0
        best = 0

        for right, current in enumerate(s):
            while current in window:
                # Remove the leftmost character until the window is valid.
                window.remove(s[left])
                left += 1

            window.add(current)
            best = max(best, right - left + 1)

        return best
